### DISCUSSION TITLE
BEテーブル設計の検討

### DISCUSSION THREAD

**Discussion by Natch** [takeyu1013/qin-todo#14] [Ideas]
# 目的
初期構想したテーブル設計をブラッシュアップする

# 初期構想
22/02/23 時点
![QinTodo drawio](https://user-images.githubusercontent.com/17216462/155288175-3bf527fc-ef22-41da-bc33-2ac342f1f5b1.svg)

**Comment by Natch** [comment]
# User

## UserId PK
よさそう。

## UserName
User という接頭辞はテーブル名から自明なので `name` でよさそう

## MailAddress
現時点で認証は Google / Apple の認証なので不要
（そもそも email を取る手段がアプリにない）

## RoleId
もしRoleに関連を持つなら、FKである必要がある
が、そもそも不要だと思う。（後述）

## CreatedAt
よさそう

## UpdatedAt
よさそう

## CreatedBy
第三者が User を作成することがそもそもできないので表現することができない。
ので不要そう。

## UpdatedBy
ユーザの情報は自身が更新することが自明なので不要そう。
また、第三者が更新することもないので。

## DeletedAt, DeletedBy, Enabled
Figma のワイヤーで「アカウントを完全に削除してよろしいですか？」と名言しており、
復元できる必要がない（論理削除である必要がない）ので不要そう。

FYI:
論理削除の仕様にした場合、論理削除した後のアプリのふるまいを別途検討する必要性がでてきて、
不要な複雑性を生みだしてしまうので避けたい。

**Reply by Natch** [reply]
supabase によって作られる可能性が大なので、
そもそも設計が必要なのかという疑問あり。

**Comment by Natch** [comment]
# TODO

## テーブル名について
TODO は1つのエンティティとして捉えることもできなくもないが、
個人的には `tasks` が良いかと思う。

なぜならテーブル（＝エンティティ）は実体そのものを表すべきだから。
TODOは暗黙的には `Task` to do だと思っていて、TODOするための「タスク」が本質的な実体だと思う。

人によってそのTODOが「タスク」なのかはわからないところだが、
 `Task` はある程度一般的に採用しても違和感がない名前かとも思う。

よりベターな命名案があればそれでも 🙆‍♂️ 

## TodoId PK 
よさそう。

## Content
よさそう

## ColumnOrder
### 名前について
接頭辞の `Column` はいらないと思う。
なぜならメタ的な情報であってTODOの実体を表すものではないから。

### 採用可否について
「並び順」に相当するものはどこかで表現する必要がありそうだが、
TODOテーブルの１カラムとして持たせるべきかはあらためて検討したほうがよさそう。

理由：
もしこのまま採用するなら、オーダーを計算するアルゴリズムをどうするか別途検討が必要。

例えば、自然数で表現するとオーダーを入れ替えたときに後続の並びを全て再計算して対象のレコードをすべて
 `UPDATE`　する必要があるためDBに対して負荷コストが高いから。

## StatusId
もし Status に関連を持つなら、FKである必要がある。
が、そもそも不要だと思う。（後述）

## UserId
もし User に関連を持つなら、FKである必要がある。

## DateId
もし Date に関連を持つなら、FKである必要がある。
が、そもそも不要だと思う。（後述）

## CreatedAt
よさそう

## UpdatedAt
よさそう

## CreatedBy
誰が作成したかは User との関連によって自明なので不要。
また、第三者が作成することもない。

## UpdatedBy
誰が更新したかは User との関連によって自明なので不要。
また、第三者が更新することもない。

## DeletedAt, DeletedBy, Enabled
論理削除は要件に含まれないので全て不要
＝　削除を取り消せるようにする必要性がなさそう
＝　物理削除でOK

理由：
一般的なTODOアプリのユースケースでは過去の日付に遡って削除したタスクを取り消す（復元）することもないし、
仮に「削除の取り消し」が要件に入ってきた場合、そのときに対応を考えれば十分。
（ちなみに、 `DeletedBy` は第三者がそもそも削除ができないので表現できない）

FYI：
物理削除の場合、レコード1件1件が User の存在有無を表すので「いつ削除したか」「誰が削除したか」といった情報は持てない。
が、問題なし。

論理削除の設計にしてしまった場合、
```
ユーザ数 (n) × 削除操作 (m) = n * m
```
で使われる予定のない無駄なレコードが無尽蔵に増えることになる。
これにより、DBの容量を無駄に圧迫し、TODOテーブルへの参照コストが上がる。

**Comment by Natch** [comment]
# Status
Todo に３つ以上の状態を持つなら必要かもしれないが、
Todo が取りうる状態は `「完了」` か `「未完了」` かの2値（boolean）で表現できる。

なので、 Todo テーブルに `boolean` 型で `done` といった名目のカラムを持たせることで必要十分。
よってテーブルまるごと不要。

---

もしくは `Todos` テーブルのカラムにする案とは別の案として、
`dones` テーブルみたいな「タスクの完了」を表すテーブルを作っても良さそう。

例えば以下のようなテーブル
```
dones
- done_id (PK)
- todo_id (FK)
- done_at
```

メリット
- ~当該Todoが完了したかどうかをレコードの有無で表現できるので nullable なカラムが不要になる~
  - ~= アプリで null かどうか気にしなくて済む~
- 完了、未完了の操作が、単に `INSERT`, `DELETE` で済むので、予期せぬ上書きが起きてしまう可能性があることを回避できる
  - （todoのカラムで表現すると、ユーザの他の操作で同じタスクに `UPDATE` したときに巻き戻ってしまう可能性がある）

## StatusId
## StatusId PK
## StatusName
## CreatedAt
## UpdatedAt
## DeletedAt
## CreatedBy
## UpdatedBy
## DeletedBy
## Enabled

**Comment by Natch** [comment]
# Date
おそらく、「今日する」「明日する」「今度する」を表すテーブルだと思われるが、
これらは Todo のもつ1属性として捉えることができる。

なので、 Todo テーブルに `date` 型の `target_date` といった名目のカラムを持たせることで必要十分。
（単に `date` だと何の日付なのかわからず、解釈の余地がでてしまうので具体的な名前にする）

よってテーブルまるごと不要。

その場合、こうした値のもたせ方ができそう。

- 今日する
  - アプリで Todo レコードを作成時、当日の日付にする
- 明日する
  - アプリで Todo レコードを作成時、明日の日付にする
- 今度する
  - アプリで Todo レコードを作成時、 `NULL` にする
  - 補足：今度する＝未定ということとして捉えられる

FYI:
ちなみに、Dateは連続性のあるスカラ値なのでテーブルに切り出すのはそもそも設計的によろしくなさそう。
（＝　正規化の対象にはなりにくい）

## DateId PK
## DateName
## CreatedAt
## UpdatedAt
## DeletedAt
## CreatedBy
## UpdatedBy
## DeletedBy
## Enabled

**Comment by Natch** [comment]
# Role
すでに会話済みだが、ユーザに役割をもたせる必要性がない（要件にない）ので、
テーブルまるごと不要。

## RoleId PK
## RoleName
## CreatedAt
## UpdatedAt
## DeletedAt
## CreatedBy
## UpdatedBy
## DeletedBy
## Enabled

**Comment by Natch** [comment]
# 全体的に言えること
↓を見ておくと良いかもです。

データベースオブジェクトの命名規約 - Qiita
https://qiita.com/genzouw/items/35022fa96c120e67c637

**Comment by Natch** [comment]
# ブラッシュアップ案1
上記の意見を踏まえて提案してみます。

```mermaid
erDiagram
    AUTHENTICATION ||--|| USER : owns
    USER ||--o{ TASK : has
    USER {
        int id
        string name
        date_time created_at
        date_time updated_at
    }
    TASK ||--o| DONE : is
    TASK {
        int id
        int  user_id
        int order
        string content
        date target_date
        date_time created_at
        date_time updated_at
    }
    DONE {
        int id
        int task_id
        date_time done_at
        date_time created_at
        date_time updated_at
    }
```

# 補足
データ型は postgres によしなに読み替えてください。

## Authentication
これは supabase で抽象化されてて、まだ詳細がわからないですが、
仕様には大きく影響がでないはずなので、いったん詳細はぼかしてます。

## User
DBに必要な情報は名前だけですが、プロフィール画像をどこに持たせるべきかは別途検討が必要そうです。
（それによってDBにも少し影響がある）

## Task
- order は要検討。
- target_date は nullable にする。もしくは別テーブルにするのもありかも。

## Done
いつ削除したかを `done_at` に持たせる。もしくは `created_at` で代用する。
（要件に「いつ削除したか」はないが、アプリの拡張性のため念の為保存しておく）

- 多重作成を防ぐため、 task_id にユニーク制約をかける


## ビジネスロジックにおけるクエリ
### タスクの一覧表示
user, task (JOIN), done (OUTER JOIN) を結合 して `SELECT`
done の有無と、 tasks.target_date でフロントの表示を変える。

### タスク追加
当該user に関連させて `task`に `INSERT`

### タスク完了
当該task に関連させて `done` に `INSERT`

- 多重実行を防ぐため、当該タスクに行ロックできるとなおよし

### タスクの完了を取り消し（未完了にする）
当該task の `done` を `DELETE`

- 多重実行を防ぐため、当該タスクに行ロックできるとなおよし

### タスクの移動（順番を入れ替える）
当該taskの order を `UPDATE`

### アカウントの編集（名前）
user の name を `UPDATE`

### アカウントの削除
当該 user, task, done をまとめて `DELETE`

・同一トランザクションで実行

**Comment by Natch** [comment]
# FYI
ER図の意味です。（ちなみに自分は正確に覚えてないです笑）
https://knooto.info/erd-crows-foot/

**Comment by Natch** [comment]
# ブラッシュアップ案2
案1の改善案です。

```mermaid
erDiagram
    AUTHENTICATION ||--|| PROFILE : owns
    PROFILE ||--o{ TASK : has
    PROFILE {
        uuid id
        text name
        timestamp created_at
        timestamp updated_at
    }
    TASK ||--o| DONE : is
    TASK ||--|| SCHEDULE : has
    TASK {
        bigint id
        bigint profile_id
        bigint order
        text content
        timestamp created_at
        timestamp updated_at
    }
    DONE {
        bigint id
        bigint task_id
        timestamp done_at
        timestamp created_at
        timestamp updated_at
    }
    SCHEDULE {
        bigint id
        bigint task_id
        date target_date
    }
```

# 変更点の補足

主な変更点は以下です。
- データ型の記述をより postgres ライクに修正
- `USER` を `PROFILE` に変更
- `TASK` の `target_date` を `SCHEDULE` テーブルに切り出して正規化


## Profile
### 名付けの件
supabase には `auth.users` テーブルというAPIでアクセスできない内部参照専用のテーブルがあります。
ただ、認証だけでなくユーザ固有の情報をAPIでアクセス可能にする必要があるため、別途テーブル定義は必要です。
しかし、 `User` というネーミングは両者を混同しかねないと判断しました。

そこで、 `Profile` というネーミングに変更しました。

> Even though Supabase provides an auth.users table, it can be helpful to create tables in the public schema for storing user data that you want to access via the API.

参考：
https://supabase.com/docs/guides/auth/managing-user-data

### PK, FKについて
同じリンクに

> When you create tables to store user data, it's helpful to reference the auth.users table in the primary key. This ensures data integrity.

とあるため、 `id` を `auth.users` に外部参照させ、 PK かつ FK にします。
（PKかつFKということをいまだかつてしたことがないので若干不安ながらも例に倣うとそうなりそうです）

## Task
target_date を nullable にする想定でしたが、null を許容することによるデメリットが経験上多いので（ケースバイケースだがアンチパターンを踏みやすくなる）テーブルに切り出してみました。

## Schedule
いつやるか（予定日）を表します。
基本的に1つしか持たないので、taskに対してユニーク制約を貼ります。

## ビジネスロジックにおけるクエリ
### タスクの一覧表示
profile, task (JOIN), done (OUTER JOIN), schedule (OUTER JOIN) を結合 して `SELECT`
done の有無と、 schedule の有無および値でフロントの表示を変える。

### タスク追加
- 当該profile に関連させて `task`に `INSERT`
- 追加対象が「今日やる」もしくは「明日やる」の場合
  - INSERT した task に関連させて `schedule` にも `INSERT`

上記を同一トランザクションで実行する

### タスク完了
当該task に関連させて `done` に `INSERT`

- 多重実行を防ぐため、当該タスクに行ロックできるとなおよし

### タスクの完了を取り消し（未完了にする）
当該task の `done` を `DELETE`

- 多重実行を防ぐため、当該タスクに行ロックできるとなおよし

### タスクの移動（順番を入れ替える）
当該taskの order を `UPDATE`

### アカウントの編集（名前）
profile の name を `UPDATE`

### アカウントの削除
当該 profile, task, done, schedule をまとめて `DELETE`

・同一トランザクションで実行

**Comment by Natch** [comment]
# ブラッシュアップ案3
案2の改善案です。

```mermaid
erDiagram
    AUTHENTICATION ||--|| PROFILE : owns
    PROFILE ||--o{ TASK : has
    PROFILE {
        uuid id
        text name
        timestamp created_at
        timestamp updated_at
    }
    TASK {
        bigint id
        bigint profile_id
        bigint order
        text content
        date scheduled_date
        boolean is_done
        timestamp done_at
        timestamp created_at
        timestamp updated_at
    }
```

# 変更点の補足

主な変更点は以下です。
- Schedule を非正規化して（テーブルを消して）Task テーブルにもたせる
- Done を非正規化して（テーブルを消して） Task テーブルにもたせる

もともとの案1, 案2 では正規化を徹底する方針でした。

しかし、バックエンドで複数のオペレーション（INSERTやUPDATE）を
同一のトランザクションとして supabese client から実行できる保証がないことがわかりました。

参考：
https://qiita.com/kabochapo/items/26b1bb753116a6904664#%E3%83%88%E3%83%A9%E3%83%B3%E3%82%B6%E3%82%AF%E3%82%B7%E3%83%A7%E3%83%B3

上記、案1, 2 は複数のオペレーションを束ねた任意のトランザクションが実行できることを前提としていたので、
もし、そうしたトランザクションを実行できなかった場合には破綻してしまいます。

また、ユニーク制約を貼るなどのバックエンドの調整、
アプリ側でのクエリの投げ分けが必要などの複雑性を引き換えにDB設計をあるべきを求めるのはコスパが悪いというのも理由です。

DB設計のあるべきを追求しておくと拡張性のメリットを得られるのですが、
今回のような知識差のあるチーム開発では難易度がかなり上がってしまうリスクのほうが大きいと判断しました。

そこで非正規化を受け入れて、簡易化をすることで極力上記のリスクを回避する方針にします。

## Task
非正規化によって以下のカラムが追加になります。

- is_done
    - boolean でもつ
- done_at
    - nullable にする
- scheduled_date
    - nullable にする

## Schedule
消して task テーブルに非正規化して持たせることにしました。

## Done
消して task テーブルに非正規化して持たせることにしました。


## ビジネスロジックにおけるクエリ

### タスクの一覧表示
profile に task を JOIN して `SELECT` するだけでOK
task.is_done の boolean の値と、 task.scheduled_date の値 ( NULL or 特定日付 ) でフロントの表示を変える。

### タスク追加
当該profile に関連させて `task`に `INSERT` これだけでOK
クライアントの操作で `INSERT` の内容が変わるだけです。

### タスク完了
当該profile に関連したすでにある `task.is_done`, `task.done_at` を `UPDATE` これだけでOK

### タスクの完了を取り消し（未完了にする）
当該profile に関連したすでにある `task.is_done`, `task.done_at` を `UPDATE` これだけでOK

### タスクの移動（順番を入れ替える）
当該profile に関連したすでにある `task.order` を `UPDATE`
（order の計算アルゴリズムは要検討。別のアプローチをとる可能性もある）

### アカウントの編集（名前）
profile の name を `UPDATE`

### アカウントの削除
当該アカウント（auth.user, profile）と task を `DELETE`

**Reply by takeyu1013** [reply]
案3がよいと思います。シンプルイズベストです。

**Comment by moto-1985** [comment]
> 案3がよいと思います。シンプルイズベストです。
同感です。オーダーは使用するライブラリーにも影響受けるのかと思いました

**Reply by Natch** [reply]
> オーダーは使用するライブラリーにも影響受けるのかと思いました

そのとおりです。ので、現状は仮になりますー。

**Comment by Natch** [comment]
案３でFIXです。

いったん現時点でFIXした図について、
参照時や修正変更があったときに管理上大変なのでgistで管理することにしました。
https://gist.github.com/Natch/e738358f1a56b78f1b49cfc92569c496
